File: GoBackend/internal/services/map_service.go
package services

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"os"

	"uber/config"
	"uber/internal/models"

	"go.mongodb.org/mongo-driver/bson"
)

// 📍 RESPONSE STRUCTS

type Coordinates struct {
	Latitude         float64 `json:"latitude"`
	Longitude        float64 `json:"longitude"`
	FormattedAddress string  `json:"formatted_address"`
}

type DistanceElement struct {
	Distance float64 `json:"distance"`
	Duration float64 `json:"duration"`
}

// 📍 GET COORDINATES
func GetAddressCoordinates(address string) (*Coordinates, error) {

	if address == "" {
		return nil, fmt.Errorf("address required")
	}

	apiKey := os.Getenv("OLA_API_KEY")

	baseURL := "https://api.olamaps.io/places/v1/geocode"

	u, err := url.Parse(baseURL)
	if err != nil {
		return nil, err
	}

	q := u.Query()
	q.Set("address", address)
	q.Set("api_key", apiKey)
	u.RawQuery = q.Encode()

	resp, err := http.Get(u.String())
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var result map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}

	results, ok := result["geocodingResults"].([]interface{})
	if !ok || len(results) == 0 {
		return nil, fmt.Errorf("no coordinates found")
	}

	first := results[0].(map[string]interface{})
	geometry := first["geometry"].(map[string]interface{})
	location := geometry["location"].(map[string]interface{})

	return &Coordinates{
		Latitude:         location["lat"].(float64),
		Longitude:        location["lng"].(float64),
		FormattedAddress: first["formatted_address"].(string),
	}, nil
}

// 🚗 GET DISTANCE + TIME
func GetDistanceAndTime(origin, destination string) (*DistanceElement, error) {

	if origin == "" || destination == "" {
		return nil, fmt.Errorf("origin and destination required")
	}

	apiKey := os.Getenv("OLA_API_KEY")

	originCoords, err := GetAddressCoordinates(origin)
	if err != nil {
		return nil, err
	}

	destCoords, err := GetAddressCoordinates(destination)
	if err != nil {
		return nil, err
	}

	baseURL := "https://api.olamaps.io/routing/v1/distanceMatrix"

	u, err := url.Parse(baseURL)
	if err != nil {
		return nil, err
	}

	q := u.Query()
	q.Set("origins", fmt.Sprintf("%f,%f", originCoords.Latitude, originCoords.Longitude))
	q.Set("destinations", fmt.Sprintf("%f,%f", destCoords.Latitude, destCoords.Longitude))
	q.Set("api_key", apiKey)
	u.RawQuery = q.Encode()

	resp, err := http.Get(u.String())
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var result map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}

	rows, ok := result["rows"].([]interface{})
	if !ok || len(rows) == 0 {
		return nil, fmt.Errorf("no distance data found")
	}

	elements := rows[0].(map[string]interface{})["elements"].([]interface{})
	data := elements[0].(map[string]interface{})

	return &DistanceElement{
		Distance: data["distance"].(float64),
		Duration: data["duration"].(float64),
	}, nil
}

// 🔍 AUTOCOMPLETE
func GetAutoCompleteSuggestions(input string) ([]interface{}, error) {

	if input == "" {
		return nil, fmt.Errorf("input required")
	}

	apiKey := os.Getenv("OLA_API_KEY")

	baseURL := "https://api.olamaps.io/places/v1/autocomplete"

	u, err := url.Parse(baseURL)
	if err != nil {
		return nil, err
	}

	q := u.Query()
	q.Set("input", input)
	q.Set("api_key", apiKey)
	u.RawQuery = q.Encode()

	resp, err := http.Get(u.String())
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var result map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}

	predictions, ok := result["predictions"].([]interface{})
	if !ok {
		return []interface{}{}, nil
	}

	return predictions, nil
}

// 📡 GET CAPTAINS IN RADIUS (FIXED)
func GetCaptainsInRadius(lat, lng, radius float64) ([]models.Captain, error) {

	filter := bson.M{
	"location": bson.M{
		"$geoWithin": bson.M{
			"$centerSphere": []interface{}{
				[]float64{lng, lat},
				radius / 6371,
			},
		},
	},
}

	cursor, err := config.DB.Collection("captains").Find(context.Background(), filter)
	if err != nil {
		return nil, err
	}

	var captains []models.Captain

	if err := cursor.All(context.Background(), &captains); err != nil {
		return nil, err
	}

	return captains, nil
}

File: SpringServer/src/main/java/com/uber/springserver/services/MapService.java
package com.uber.springserver.services;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.uber.springserver.models.Captain;
import com.uber.springserver.repositories.CaptainRepository;
import com.uber.springserver.utils.EnvUtil;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Service
public class MapService {

    private final RestTemplate restTemplate = new RestTemplate();
    private final ObjectMapper objectMapper = new ObjectMapper();
    private final CaptainRepository captainRepository;

    public MapService(CaptainRepository captainRepository) {
        this.captainRepository = captainRepository;
    }

    public Map<String, Object> getAddressCoordinates(String address) throws Exception {
        if (address == null || address.isBlank()) {
            throw new IllegalArgumentException("address required");
        }

        String apiKey = EnvUtil.get("OLA_API_KEY", "");
        String url = "https://api.olamaps.io/places/v1/geocode?address=" +
                URLEncoder.encode(address, StandardCharsets.UTF_8) +
                "&api_key=" + URLEncoder.encode(apiKey, StandardCharsets.UTF_8);

        ResponseEntity<String> response = restTemplate.getForEntity(url, String.class);
        JsonNode root = objectMapper.readTree(response.getBody());
        JsonNode results = root.path("geocodingResults");
        if (!results.isArray() || results.isEmpty()) {
            throw new IllegalStateException("no coordinates found");
        }

        JsonNode first = results.get(0);
        JsonNode location = first.path("geometry").path("location");

        Map<String, Object> data = new HashMap<>();
        data.put("latitude", location.path("lat").asDouble());
        data.put("longitude", location.path("lng").asDouble());
        data.put("formatted_address", first.path("formatted_address").asText());
        return data;
    }

    public Map<String, Object> getDistanceAndTime(String origin, String destination) throws Exception {
        if (origin == null || destination == null || origin.isBlank() || destination.isBlank()) {
            throw new IllegalArgumentException("origin and destination required");
        }

        Map<String, Object> originCoords = getAddressCoordinates(origin);
        Map<String, Object> destinationCoords = getAddressCoordinates(destination);

        String apiKey = EnvUtil.get("OLA_API_KEY", "");
        String origins = originCoords.get("latitude") + "," + originCoords.get("longitude");
        String destinations = destinationCoords.get("latitude") + "," + destinationCoords.get("longitude");

        String url = "https://api.olamaps.io/routing/v1/distanceMatrix?origins=" +
                URLEncoder.encode(origins, StandardCharsets.UTF_8) +
                "&destinations=" + URLEncoder.encode(destinations, StandardCharsets.UTF_8) +
                "&api_key=" + URLEncoder.encode(apiKey, StandardCharsets.UTF_8);

        double distance = 0;
        double duration = 0;

        try {
            ResponseEntity<String> response = restTemplate.getForEntity(url, String.class);
            JsonNode root = objectMapper.readTree(response.getBody());

            JsonNode rows = root.path("rows");
            if (rows.isArray() && !rows.isEmpty()) {
                JsonNode elements = rows.get(0).path("elements");
                if (elements.isArray() && !elements.isEmpty()) {
                    JsonNode firstElement = elements.get(0);

                    JsonNode distanceNode = firstElement.path("distance");
                    JsonNode durationNode = firstElement.path("duration");

                    distance = readNodeValue(distanceNode, "distance", "distanceMeters", "value");
                    duration = readNodeValue(durationNode, "duration", "durationSeconds", "value");
                }
            }
        } catch (Exception ignored) {
            // Geocoding succeeded, so calculate a real fallback distance below.
        }

        if (distance <= 0) {
            double originLat = ((Number) originCoords.get("latitude")).doubleValue();
            double originLng = ((Number) originCoords.get("longitude")).doubleValue();
            double destinationLat = ((Number) destinationCoords.get("latitude")).doubleValue();
            double destinationLng = ((Number) destinationCoords.get("longitude")).doubleValue();

            double directDistanceKm = haversineKm(originLat, originLng, destinationLat, destinationLng);
            distance = directDistanceKm * 1000;
            if (duration <= 0) {
                duration = directDistanceKm > 0 ? directDistanceKm * 120 : 0;
            }
        }

        Map<String, Object> data = new HashMap<>();
        data.put("distance", distance);
        data.put("duration", duration);
        return data;
    }

    private double readNodeValue(JsonNode node, String... fieldNames) {
        if (node == null || node.isMissingNode() || node.isNull()) {
            return 0;
        }
        if (node.isObject()) {
            for (String fieldName : fieldNames) {
                JsonNode valueNode = node.path(fieldName);
                if (valueNode.isNumber()) {
                    return valueNode.asDouble();
                }
            }
        }
        return node.asDouble(0);
    }

    private double haversineKm(double lat1, double lng1, double lat2, double lng2) {
        final double earthRadiusKm = 6371.0;
        double dLat = Math.toRadians(lat2 - lat1);
        double dLng = Math.toRadians(lng2 - lng1);
        double a = Math.sin(dLat / 2) * Math.sin(dLat / 2)
                + Math.cos(Math.toRadians(lat1)) * Math.cos(Math.toRadians(lat2))
                * Math.sin(dLng / 2) * Math.sin(dLng / 2);
        double c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
        return earthRadiusKm * c;
    }

    public List<Object> getAutoCompleteSuggestions(String input) throws Exception {
        if (input == null || input.isBlank()) {
            throw new IllegalArgumentException("input required");
        }

        String apiKey = EnvUtil.get("OLA_API_KEY", "");
        String url = "https://api.olamaps.io/places/v1/autocomplete?input=" +
                URLEncoder.encode(input, StandardCharsets.UTF_8) +
                "&api_key=" + URLEncoder.encode(apiKey, StandardCharsets.UTF_8);

        ResponseEntity<String> response = restTemplate.getForEntity(url, String.class);
        JsonNode root = objectMapper.readTree(response.getBody());

        List<Object> predictions = new ArrayList<>();
        JsonNode predictionNode = root.path("predictions");
        if (predictionNode.isArray()) {
            for (JsonNode node : predictionNode) {
                predictions.add(objectMapper.convertValue(node, Object.class));
            }
        }

        return predictions;
    }

    public List<Captain> getCaptainsInRadius(double lat, double lng, double radiusKm) {
        double latDelta = radiusKm / 111.0;
        double lngDelta = radiusKm / (111.0 * Math.max(Math.cos(Math.toRadians(lat)), 0.1));

        List<Captain> captains = captainRepository.findAll();
        List<Captain> inRadius = new ArrayList<>();
        for (Captain captain : captains) {
            if (captain.getLocation() == null || captain.getLocation().getLtd() == null || captain.getLocation().getLng() == null) {
                continue;
            }
            double cLat = captain.getLocation().getLtd();
            double cLng = captain.getLocation().getLng();
            if (cLat >= lat - latDelta && cLat <= lat + latDelta && cLng >= lng - lngDelta && cLng <= lng + lngDelta) {
                inRadius.add(captain);
            }
        }
        return inRadius;
    }
}
